internal/charmstore/resources.go

```diff
@@ -0,0 +1,256 @@
+	for name, meta := range entity.CharmMeta.Resources {
+		doc, err := s.latestResource(baseEntity, channel, name)
+		if errgo.Cause(err) == resourceNotFound {
+			// Convert the "not found" error into placeholder. We do
+			// this instead of pre-populating the database with
+			// placeholders. This keeps the machinery simpler while
+			// still correct.
+			doc = &mongodoc.Resource{
+				BaseURL: baseEntity.URL,
+				Name:    meta.Name,
+				// Revision, Fingerprint, etc. are not set.
+			}
+		} else if err != nil {
+			return nil, errgo.Notef(err, "failed to get resource %q", name)
+		}
+		docs = append(docs, doc)
+	}
+	sortResources(docs)
+	return docs, nil
+}
+
+// ResourceBlob holds the information specific to a single resource blob.
+type ResourceBlob struct {
+	io.Reader
+
+	// Hash is the SHA-384 checksum of the blob.
+	Hash string
+
+	// Size is the size of the blob in bytes.
+	Size int64
+}
+
+// TODO(ericsnow) We will need Store.nextResourceRevision() to get the
+// value to pass to addResource().
+
+func (s *Store) addResource(entity *mongodoc.Entity, name string, blob ResourceBlob, newRevision int) error {
+	if !charmHasResource(entity.CharmMeta, name) {
+		return errgo.Newf("charm does not have resource %q", name)
+	}
+
+	blobName, err := s.storeResource(blob)
+	if err != nil {
+		return errgo.Mask(err, errgo.Is(resourceNotFound))
+	}
+
+	doc := &mongodoc.Resource{
+		BaseURL:    entity.BaseURL,
+		Name:       name,
+		Revision:   newRevision,
+		BlobHash:   blob.Hash,
+		Size:       blob.Size,
+		BlobName:   blobName,
+		UploadTime: time.Now().UTC(),
+	}
+	if s.insertResource(doc); err != nil {
+		if err := s.BlobStore.Remove(blobName); err != nil {
+			logger.Errorf("cannot remove blob %s after error: %v", blobName, err)
+		}
+		return errgo.Mask(err)
+	}
+	return nil
+}
+
+func (s *Store) insertResource(doc *mongodoc.Resource) error {
+	if err := doc.Validate(); err != nil {
+		return errgo.Mask(err)
+	}
+	err := s.DB.Resources().Insert(doc)
+	if err != nil {
+		if mgo.IsDup(err) {
+			return errgo.WithCausef(nil, params.ErrDuplicateUpload, "")
+		}
+		return errgo.Notef(err, "cannot insert resource")
+	}
+	return nil
+}
+
+func (s *Store) storeResource(blob ResourceBlob) (string, error) {
+	name := bson.NewObjectId().Hex()
+	// TODO(ericsnow) We will finish this in a follow-up patch.
+	return name, nil
+}
+
+func (s *Store) setResource(entity *mongodoc.Entity, channel params.Channel, name string, revision int) error {
+	if channel == params.NoChannel {
+		return errgo.New("missing channel")
+	}
+	if channel == params.UnpublishedChannel {
+		return errgo.New("cannot publish to unpublished channel")
+	}
+
+	doc, err := s.resource(entity.BaseURL, name, revision)
+	if err != nil {
+		return errgo.Mask(err, errgo.Is(resourceNotFound))
+	}
+
+	if !charmHasResource(entity.CharmMeta, doc.Name) {
+		return errgo.Mask(err)
+	}
+
+	update := bson.D{{fmt.Sprintf("channelresources.%s.%s", channel, name), revision}}
```

Comment:
Looks like this could be a problem if resource names might contain dots or other nasty characters.

You'll have to either permanently exclude nasty characters from the resource name, encode resource names, or use a slice in the doc, I think.